File: 12_path_in_matrx/main.cpp
#include <iostream>
#include <vector>

bool HasPathCore(const std::vector<std::vector<char>>& matrix, const std::string& str,
                 int rows, int cols, int row, int col, int& path_len, std::vector<bool>& visited) {
    if (path_len == str.length()) return true;
    bool has_path = false;
    if (row >= 0 && row < rows && col >= 0 && col < cols &&
        matrix[row][col] == str[path_len] && !visited[row * cols + col]) {
        ++path_len;
        visited[row * cols + col] = true;
        has_path = HasPathCore(matrix, str, rows, cols, row, col - 1, path_len, visited) ||
            HasPathCore(matrix, str, rows, cols, row - 1, col, path_len, visited) ||
            HasPathCore(matrix, str, rows, cols, row, col + 1, path_len, visited) ||
            HasPathCore(matrix, str, rows, cols, row + 1, col, path_len, visited);
        if (!has_path) {
            --path_len;
            visited[row * cols + col] = false;
        } else std::cout << matrix[row][col];
    }
    return has_path;
}

bool HasPath(const std::vector<std::vector<char>>& matrix, const std::string& str) {
    if (matrix.size() == 0 || matrix[0].size() == 0 || str.length() == 0) return false;

    int rows = matrix.size();
    int cols = matrix[0].size();
    std::vector<bool> visited(rows * cols);
    int path_len = 0;

    for (int i = 0; i < rows; ++i) {
        for (int j = 0; j < rows; ++j) {
            if (HasPathCore(matrix, str, rows, cols, i, j, path_len, visited)) {
                return true;
            }
        }
    }
    
    return false;
}

int main() {
    std::vector<std::vector<char>> matrix = {
        {'a', 'b', 't', 'g'},
        {'c', 'f', 'c', 's'},
        {'j', 'd', 'e', 'h'}
    };
    bool has_path = HasPath(matrix, "bfce");
    std::cout << std::endl << has_path << std::endl;
}

File: 12_path_in_matrx/go/main.go
package main

import (
	"fmt"
)

func hasPathCore(matrix [][]byte, rows, cols, row, col int,
	s string, pathLen *int, visited *[]bool) bool {
	if *pathLen == len(s) {
		return true
	}
	hasPath := false
	if row >= 0 && row < rows && col >= 0 && col < cols &&
		matrix[row][col] == s[*pathLen] && !(*visited)[row * cols + col] {
		*pathLen++
		(*visited)[row * cols + col] = true
		hasPath = hasPathCore(matrix, rows, cols, row, col - 1, s, pathLen, visited) ||
			hasPathCore(matrix, rows, cols, row - 1, col, s, pathLen, visited) ||
			hasPathCore(matrix, rows, cols, row, col + 1, s, pathLen, visited) ||
			hasPathCore(matrix, rows, cols, row + 1, col, s, pathLen, visited)
		if !hasPath {
			*pathLen--
			(*visited)[row * cols + col] = false
		}
	}
	return hasPath
}

func HasPath(matrix [][]byte, s string) bool {
	if matrix == nil || len(matrix) == 0 || len(matrix[0]) == 0 || s == "" {
		return false
	}
	rows, cols := len(matrix), len(matrix[0])
	pathLen, visited := 0, make([]bool, rows * cols)
	for i := 0; i < rows; i++ {
		for j := 0; j < cols; j++ {
			if hasPathCore(matrix, rows, cols, i, j, s, &pathLen, &visited) {
				return true
			}
		}
	}
	return false
}

func main() {
	matrix := [][]byte{
		{'a', 'b', 't', 'g'},
		{'c', 'f', 'c', 's'},
		{'j', 'd', 'e', 'h'},
	}
	s := "bfdehsc"
	fmt.Println(HasPath(matrix, s))
}
